app/src/main/java/nodomain/freeyourgadget/gadgetbridge/service/devices/hplus/HPlusSupport.java

```diff
@@ -0,0 +1,881 @@
+            } else {
+                message = title;
+                for (int i = message.length(); i < 12; i++)
+                    message += " ";
+            }
+        }
+        message += body;
+
+        int length = message.length() / 17;
+
+        builder.write(ctrlCharacteristic, new byte[]{HPlusConstants.COMMAND_ACTION_INCOMING_SOCIAL, (byte) 255});
+
+        int remaining = 0;
+
+        if (message.length() % 17 > 0)
+            remaining = length + 1;
+        else
+            remaining = length;
+
+        msg[1] = (byte) remaining;
+        int message_index = 0;
+        int i = 3;
+
+        for (int j = 0; j < message.length(); j++) {
+            msg[i++] = (byte) message.charAt(j);
+
+            if (i == msg.length) {
+                message_index++;
+                msg[2] = (byte) message_index;
+                builder.write(ctrlCharacteristic, msg);
+
+                msg = msg.clone();
+                for (i = 3; i < msg.length; i++)
+                    msg[i] = 32;
+
+                if (message_index < remaining)
+                    i = 3;
+                else
+                    break;
+            }
+        }
+
+        msg[2] = (byte) remaining;
+
+        builder.write(ctrlCharacteristic, msg);
+        builder.queue(getQueue());
+    }
+
+    public boolean isExpectedDevice(BluetoothDevice device) {
+        return true;
+    }
+
+    public void close() {
+    }
+
+    @Override
+    public boolean onCharacteristicChanged(BluetoothGatt gatt,
+                                           BluetoothGattCharacteristic characteristic) {
+        if (super.onCharacteristicChanged(gatt, characteristic)) {
+            return true;
+        }
+
+        UUID characteristicUUID = characteristic.getUuid();
+        byte[] data = characteristic.getValue();
+        if (data.length == 0)
+            return true;
+
+        switch (data[0]) {
+            case HPlusConstants.DATA_STATS:
+                return processDataStats(data);
+            case HPlusConstants.DATA_SLEEP:
+                return processSleepStats(data);
+            default:
+                LOG.info("Unhandled characteristic changed: " + characteristicUUID);
+
+        }
+        return false;
+    }
+
+    private boolean processSleepStats(byte[] data) {
+        LOG.debug("Process Sleep Stats");
+
+        if (data.length < 19) {
+            LOG.error("Invalid Sleep Message Length " + data.length);
+            return false;
+        }
+        HPlusSleepRecord record = new HPlusSleepRecord(data);
+
+        try (DBHandler handler = GBApplication.acquireDB()) {
+            DaoSession session = handler.getDaoSession();
+
+            Device device = DBHelper.getDevice(getDevice(), session);
+            User user = DBHelper.getUser(session);
+            int ts = (int) (System.currentTimeMillis() / 1000);
+            HPlusSampleProvider provider = new HPlusSampleProvider(gbDevice, session);
+
+            //TODO: Store Sample. How?
```

Comment:
This shouldn't be a problem. You have total freedom over how you store sample data. Your custom SampleProvider can then figure out the details to provide the samples in a "normalized" way so that the charts can display them.

E.g. some Pebble Health data is stored as some kind of "overlay" data and combined with the other data by the sample provider.

However, if you realize that this does not make sense for HPlus, we can also provide some kind of aggregate data interface. Needs some thinking because we don't have something like that yet.